src/util/codec/mysql/json/json_modify.rs

```diff
@@ -0,0 +1,175 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use super::Json;
+use super::super::Result;
+use super::path_expr::{PathLeg, PathExpression, contains_any_asterisk};
+
+/// `ModifyType` is for modify a JSON.
+#[derive(Clone, Debug)]
+pub enum ModifyType {
+    /// `Insert` is for inserting a new element into a JSON.
+    Insert,
+    /// `Replace` is for replacing a old element from a JSON.
+    Replace,
+    /// `Set` = `Insert` | `Replace`
+    Set,
+}
+
+impl Json {
+    // Modifies a Json object by insert, replace or set.
+    // All path expressions cannot contain * or ** wildcard.
+    // If any error occurs, the input won't be changed.
+    pub fn modify(self,
+                  path_expr_list: &[PathExpression],
+                  mut values: Vec<Json>,
+                  mt: ModifyType)
+                  -> (Json, Result<()>) {
+        if path_expr_list.len() != values.len() {
+            return (self, Err(box_err!("Incorrect parameter count")));
+        }
+        for expr in path_expr_list {
+            if contains_any_asterisk(expr.flags) {
+                return (self, Err(box_err!("Invalid path expression")));
+            }
+        }
+        let mut j = self;
+        for (expr, value) in path_expr_list.iter().zip(values.drain(..)) {
+            j = set_json(j, &expr.legs, value, mt.clone());
+        }
+        (j, Ok(()))
+    }
+}
+
+// `set_json` is used in Json::modify().
+fn set_json(j: Json, path_legs: &[PathLeg], value: Json, mt: ModifyType) -> Json {
+    if path_legs.is_empty() {
+        match mt {
+            ModifyType::Replace | ModifyType::Set => return value,
+            _ => return j,
+        }
+    }
+    let (current_leg, sub_path_legs) = (&path_legs[0], &path_legs[1..]);
+    match *current_leg {
+        PathLeg::Index(i) => {
+            let index = i as usize;
+            // If `j` is not an array, we should autowrap it to be an array.
+            // Then if the length of result array equals to 1, it's unwraped.
+            let (mut array, wrapped) = match j {
+                Json::Array(array) => (array, false),
+                _ => (vec![j], true),
+            };
+            if array.len() > index {
+                // e.g. json_replace('[1, 2, 3]', '$[0]', "x") => '["x", 2, 3]'
+                if index == (array.len() - 1) {
+                    let chosen = array.pop().unwrap();
+                    array.push(set_json(chosen, sub_path_legs, value, mt));
+                } else {
+                    let mut right = array.split_off(index + 1);
+                    let chosen = array.pop().unwrap();
+                    array.push(set_json(chosen, sub_path_legs, value, mt));
+                    array.append(&mut right);
```

Comment:
Here `split_off` https://doc.rust-lang.org/src/collections/vec.rs.html#1201 and `append` may cost too much, could we adjust `set_json `'s input, so we could use `array.get_mut(index)` instead here?